tutorials/haskell/gui-application/tutorial.md

```diff
@@ -0,0 +1,825 @@
+cases are most frequent:
+
+* Just `IO ()`, no information is given to the handler it is not expected to
+  return anything. Example of such signal is `showSignal`.
+
+* Handlers that are given arguments: `a -> IO Bool`. Example of such signal
+  is `focus` whose handlers have the type `DirectionType -> IO Bool`.
+  Another interesting thing here is returned value of the type `Bool`. This
+  is a convention in GTK+ allowing to disable default handling of some
+  signals. If we return `True`, default handling will be disabled, while
+  `False` will keep it active executing our handler *and* default handler as
+  well.
+
+There is one more way to get some information from within signal's handler.
+Some signals dictate that handler should live in special monad called
+`EventM` instead of plain `IO`. Signals that like their handlers to be in
+`EventM` are called events.
+
+What is the `EventM` monad? Actually it's a type synonym for a simple monad
+stack with `IO` at the bottom:
+
+```haskell
+type EventM t = ReaderT (Ptr t) IO
+```
+
+This is just a reader monad transformer on top of `IO`. `t` specifies type
+of information we can extract and which helper function we can use inside
+the `EventM` monad. These are different for every event. For example
+`configureEvent` allows to extract information about window size, while
+`keyPressEvent` event provides information about key that has been pressed,
+which modifier key was held at that time and so forth. The type system does
+not allow to try to extract information that particular event does not
+provide.
+
+I would like to quote the docs to accent importance of returned Boolean
+value:
+
+> Note that an event handler must always returns `True` if the event was
+> handled or `False` if the event should be dealt with by another event
+> handler. For instance, a handler for a key press should return `False` if
+> the pressed key is not one of those that the widget reacts to. In this
+> case the event is passed to the parent widgets. This ensures that
+> pressing, say, `Alt-F` opens the file menu even if the current input focus
+> is in a text entry widget. In order to facilitate writing handlers that
+> may abort handling an event, this module provides the function `tryEvent`.
+> This function catches pattern match exceptions and returns `False`. If the
+> signal successfully runs to its end, it returns `True`.
+
+Knowing all that, we can write a simple handler to run on button activation.
+Looking at the “signals” section in `Graphics.UI.Gtk.Buttons.Button`,
+`buttonActivated` looks like our frined here:
+
+```haskell
+-- | Emitted when the button has been activated (pressed and released).
+buttonActivated :: ButtonClass self => Signal self (IO ())
+```
+
+Just for a test, let's re-write `mkBtn` to attach a handler that will update
+the display with name of pressed button (we still don't know the whole lot
+to make a working calculator):
```

Comment:
@mrkkrp 

- "that will update the display with **the** name of **the** pressed button".
- "we still don't know **a**~~the~~ whole lot".